source/agora/network/Crawler.d

```diff
@@ -0,0 +1,526 @@
+                        this.discovered_node_list.insertBack(network_client_holder);
+                    }
+                    else
+                        discovered_node_map.remove(client_address);
+
+                    crawl_result_holder.crawl_results.remove(client_address);
+                }
+                else
+                    this.discovered_node_list.insertBack(network_client_holder);
+
+                // Crawling failed for this node, and should be continued immediately
+                // without waiting
+                continue;
+            }
+
+            // Do simple checks on the recommended nodes, and if checks pass,
+            // then add them to the data structures
+            this.processNetworkAddresses(client_address, node_info.addresses.pickRandom(10));
+
+            // Node proved that, it implements getNodeInfo method, and it is still
+            // running, so it's statistics should be added to the crawling results
+            if (auto it = client_address in this.crawl_result_holder)
+                it.crawl_time = this.clock.networkTime();
+            else
+                this.crawl_result_holder[client_address] =
+                    this.populateCrawlResult(
+                        this.node_locator.extractValues(client_address,
+                            [
+                                "continent->names->en",
+                                "country->names->en",
+                                "city->names->en",
+                                "location->latitude",
+                                "location->longitude",
+                            ]) ~ node_info.os);
+
+            // After a succesfull crawl, wait before starting a new one
+            this.taskman.wait(this.config.node.crawling_interval);
+            this.discovered_node_list.insertBack(network_client_holder);
+        }
+
+        // After all the fibers exited the for loop, and there is no fiber
+        // trying to use the NodeLocator, resources can be freed
+        if (this.already_shutdown_cnt++ == this.config.node.num_of_crawlers)
+            do_stop();
+    }
+
+    /***************************************************************************
+
+        Populates a `CrawlResult` based on the string values passed in and
+        based on current time.
+
+        Params:
+            extracted_values = the values that will be used to populate the
+            `CrawlResult` object
+
+        Returns:
+            a `CrawlResult` object populated from `extracted_values`
+
+    ***************************************************************************/
+
+    private CrawlResult populateCrawlResult (string[] extracted_values) @safe nothrow
+    {
+        CrawlResult crawl_result;
+        foreach (i, ref field; crawl_result.tupleof)
+            static if (is(typeof(field) == string))
+                field = extracted_values[i];
+            else if (is(typeof(field) == TimePoint))
```

Comment:
`else static if`

Otherwise it's a runtime `if`. It's one of those D's gotcha.
Also what happens for other fields ? We just ignore it ? I think we should have a `static assert` for those.